text/0000-intra-rustdoc-links.md

```diff
@@ -0,0 +1,465 @@
+/// [`postfix_predicate!`] instead of implementing this directly.
+pub trait Expression {
+    type SqlType;
+}
+
+/// Describes how a type can be represented as an expression for a given type.
+/// These types couldn't just implement [`Expression`] directly, as many things
+/// can be used as an expression of multiple types. ([`String`] for example, can
+/// be used as either [`VarChar`] or [`Text`]).
+///
+/// [`VarChar`]: diesel::types::VarChar
+/// [`Text`]: diesel::types::Text
+pub trait AsExpression<T> {
+    type Expression: Expression<SqlType=T>;
+    fn as_expression(self) -> Self::Expression;
+}
+```
+
+Please note:
+
+- This uses implied shortcut reference links most often.
+  Since the original documentation put the type/trait names in back ticks to render them as code, we preserved this style.
+  (We don't propose this as a general convention, though.)
+- Even though implied shortcut reference links could be used throughout,
+  they are not used for the last two links (to `VarChar` and `Text`),
+  which are not in scope and need to be linked to by their absolute Rust path.
+  To make reading easier and less noisy, reference links are used to rename the links.
+  (An assumption is that most readers will recognize these names and know they are part of `diesel::types`.)
+
+
+# How We Teach This
+[how-we-teach-this]: #how-we-teach-this
+
+- Extend the documentation chapter of the book with a subchapter on How to Link to Items.
+- Reference the chapter on the module system, to let reads familiarize themselves with Rust paths.
+- Maybe present an example use case of a module whose documentation links to several related items.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Rustdoc gets more complex.
+- These links won't work when the doc comments are rendered with a default Markdown renderer.
+- The Rust paths might conflict with other valid links,
+  though we could not think of any.
+
+
+# Possible Extensions
+[possible-extensions]: #possible-extensions
+
+## Linking to fields and variants
+
+To link to the fields of a `struct`
+we propose to write the path to the struct,
+followed by a dot, followed by the field name.
+
+For example:
+
+```markdown
+This is stored in the [`size`](storage::Filesystem.size) field.
+```
+
+To link to the variants of an `enum`,
+we propose to write the path to the enum,
+followed by two colons, followed by the field name,
+just like `use Foo::Bar` can be used to import the `Bar` variant of an `enum Foo`.
+
+For example:
+
+```markdown
+For custom settings, supply the [`Custom`](storage::Engine::Other) field.
+```
+
+## Linking to associated items
+
+To link to associated items,
+i.e., the associated functions, types, and constants of a trait,
+it may be necessary to use fully-qualified paths,
+like `See the [<Foo as Bar>::bar()] method`.
+We have yet to analyze in which cases this is necessary,
+and this syntax is currently not described in [the reference's section on paths][ref-paths].
```

Comment:
While it's not in the reference under paths, thanks to [rust-lang-nursery/reference#45](https://github.com/rust-lang-nursery/reference/pull/45), there's now [a mention of fully-qualified paths under expressions](https://github.com/rust-lang-nursery/reference/blob/d9243090c118439c69d325fc0509938809142666/src/expressions.md#disambiguating-function-calls)!